### PR TITLE
Refactor temporal extract date part kernels

### LINKED ISSUE #5266: Temporal Extract/Date Part Kernel
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently we provide a mix of kernels for extracting portions of temporal types, https://docs.rs/arrow-arith/latest/arrow_arith/temporal/index.html

There are a couple of limitations with this implementation:

* The use of generics makes it non-trivial to support types such as Time32/Time64 (#5261) or intervals (#TBD)
* The way the kernels are implemented results in redundant code generation for impossible combinations, e.g. a Date32Array with a DataType of Timestamp
* The use of as_datetime and similar to always proxy via chrono is not only very inefficient, but also hard to follow
* They make use of builders instead of the faster (and more concise) `PrimitiveArray::try_unary` methods
* The code in general is quite hard to follow

**Describe the solution you'd like**

I would like to take a similar approach to used in #4465, where we have a single method that accepts an operation and array, and then dispatches the operation to the specialized implementation for that particular array type. Aside from simplifying the implementation, we could also expose this publicly to facilitate implementation of the SQL extract/date_part function - https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT.

In particular I would like something akin to

```
pub enum DatePart {
    Century,
    Year,
    ...
    Minute,
    ...
    DayOfYear,
    DayOfWeek,
}

pub fn date_part(array: &dyn Array, part: DatePart) -> Result<Int32Array> {
    match array.data_type() {
        ...
    }
}
```

The existing kernels would be deprecated and updated to act as simple shims to this dynamic dispatch logic, as has been done with the previous datum migrations.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #5266

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

No, aside from new function. Existing functions (`hour(...)`, `year(...)`) are unchanged, only internals are changed and should maintain existing behaviour.

### CODE REVIEW

File: arrow-arith/src/temporal.rs [hunk 1]
```diff
@@ -19,105 +19,352 @@
...
-            },
-            None => builder.append_null(),
+fn date_part_time32_s(
+    array: &PrimitiveArray<Time32SecondType>,
+    part: DatePart,
+) -> Result<Int32Array, ArrowError> {
+    match part {
+        DatePart::Hour => Ok(array.unary_opt(|d| time32s_to_time(d).map(|c| c.hour() as i32))),
+        // TODO expand support for Time types, see: https://github.com/apache/arrow-rs/issues/5261
+        _ => return_compute_error_with!(format!("{part} does not support"), array.data_type()),
```
Comment (Jefffrey) [inline arrow-arith/src/temporal.rs:122]: Can tackle this as separate issue to make this refactoring simpler (maintain existing behaviour)

**Review by Jefffrey — COMMENTED** [review]
Let me know what you think @tustvold , if this falls in line with what you're imagining.

I tried to make things simpler, without resorting to macros so the code might be duplicated a bit but I tried to reduce as much as I could.

Happy for any feedback

File: arrow-arith/src/temporal.rs [hunk 2]
```diff
@@ -19,105 +19,352 @@
...
-                Some(dt) => builder.append_value(op(dt)),
-                None => builder.append_null(),
-            }
-        } else {
-            builder.append_null();
-        }
-    });
+/// Returns function to extract relevant [`DatePart`] from a [`NaiveDateTime`].
+fn get_naive_date_time_part_extract_fn(part: DatePart) -> fn(NaiveDateTime) -> i32 {
+    match part {
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:77]: Perhaps we could use https://docs.rs/chrono/latest/chrono/trait.Datelike.html and https://docs.rs/chrono/latest/chrono/trait.Timelike.html to reduce duplication with the method below

File: arrow-arith/src/temporal.rs [hunk 3]
```diff
@@ -19,105 +19,352 @@
...
-            match as_datetime::<T>(i64::from(value)) {
-                Some(dt) => builder.append_value(op(dt)),
-                None => builder.append_null(),
-            }
-        } else {
-            builder.append_null();
-        }
-    });
+/// Returns function to extract relevant [`DatePart`] from a [`NaiveDateTime`].
+fn get_naive_date_time_part_extract_fn(part: DatePart) -> fn(NaiveDateTime) -> i32 {
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:76]: I'm curious why this doesn't need to be `impl fn`

File: arrow-arith/src/temporal.rs [hunk 4]
```diff
@@ -19,105 +19,352 @@
...
+/// Dispatch to specialized function depending on the array type. Since we don't need to consider
+/// dictionary arrays here, can strictly control return type to be [`Int32Array`].
+fn primitive_array_date_part<T: ArrowPrimitiveType>(
+    array: &PrimitiveArray<T>,
+    part: DatePart,
+) -> Result<Int32Array, ArrowError> {
+    match array.data_type() {
+        DataType::Date32 => {
+            let array = downcast_array::<Date32Array>(array);
+            Ok(date_part_date32(&array, part))
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:306]:
```suggestion
            Ok(date_part_date32(array.as_primitive::<Date32Type>(), part))
```

I prefer the `AsArray` formulation

File: arrow-arith/src/temporal.rs [hunk 5]
```diff
@@ -19,105 +19,352 @@
...
+) -> Int32Array {
+    let map_func = get_naive_date_time_part_extract_fn(part);
+    array.unary_opt(|d| timestamp_ns_to_datetime(d).map(map_func))
+}
+
+fn date_part_timestamp_ns_tz(
+    array: &PrimitiveArray<TimestampNanosecondType>,
+    part: DatePart,
+    tz: Tz,
+) -> Int32Array {
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:265]: One way that you might be able to avoid this duplication would be to make the functions take a generic `Timezone` argument, and then provide `Utc` as the timezone when none is provided

**Review by tustvold — APPROVED** [review]
I think this is a nice step forward, I have some ideas on how we might use generic to reduce some of the duplication, but I'd need to try it out to know if it works :sweat_smile:

File: arrow-arith/src/temporal.rs (hunk 3 above)
Comment (Jefffrey) [reply arrow-arith/src/temporal.rs:76]: No reason, really. I just stuck with the first signature that worked here :sweat_smile:

**Comment by Jefffrey** [comment]
> I think this is a nice step forward, I have some ideas on how we might use generic to reduce some of the duplication, but I'd need to try it out to know if it works 😅

I've refactored it to use generics more, let me know if it looks better this way :+1:

File: arrow-arith/src/temporal.rs [hunk 6]
```diff
@@ -19,105 +19,273 @@
...
+fn get_tz(dt: &DataType) -> Result<Option<Tz>, ArrowError> {
+    match dt {
+        DataType::Timestamp(_, Some(tz)) => Ok(Some(tz.parse::<Tz>()?)),
+        DataType::Timestamp(_, None) => Ok(None),
+        _ => Err(ArrowError::CastError(format!("Not a timestamp type: {dt}"))),
+    }
+}
+
+/// Implement the specialized functions for extracting date part from temporal arrays.
+trait ExtractDatePartExt {
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:150]: Nice, this is exactly what I was thinking of :+1:

**Review by tustvold — APPROVED** [review]
This is very nice, thank you

File: arrow-arith/src/temporal.rs [hunk 7]
```diff
@@ -19,105 +19,273 @@
...
+    fn date_part(&self, part: DatePart) -> Result<Int32Array, ArrowError> {
+        // Date32 only encodes number of days, so these will always be 0
+        if let DatePart::Hour
+        | DatePart::Minute
+        | DatePart::Second
+        | DatePart::Millisecond
+        | DatePart::Microsecond
+        | DatePart::Nanosecond = part
+        {
+            Ok(self.unary(|_| 0))
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:204]: FWIW it might be cheaper to do `Int32Array::new(vec![0; self.len()].into(), self.nulls().clone())` or something

File: arrow-arith/src/temporal.rs [hunk 8]
```diff
@@ -228,14 +372,14 @@ where
...
     i64: From<T::Native>,
 {
-    time_fraction_internal(array, "year", |t| t.year())
+    date_part_primitive(array, DatePart::Year)
 }
 
 /// Extracts the quarter of a given temporal array as an array of integersa within
 /// the range of [1, 4]. If the given array isn't temporal primitive or dictionary array,
 /// an `Err` will be returned.
 pub fn quarter_dyn(array: &dyn Array) -> Result<ArrayRef, ArrowError> {
```
Comment (tustvold) [inline arrow-arith/src/temporal.rs:381]: Should we deprecate these functions?

File: arrow-arith/src/temporal.rs (hunk 7 above)
Comment (Jefffrey) [reply arrow-arith/src/temporal.rs]: Good point, applied change

File: arrow-arith/src/temporal.rs (hunk 8 above)
Comment (Jefffrey) [reply arrow-arith/src/temporal.rs:384]: Probably for the best, as if in the future we want to support more parts like `century` or `decade` in arrow-rs natively, we wouldn't have to add these shim functions for consistency if they are already deprecated.

That does raise a question that comes to mind, that if we add more of these parts, by extending the `DatePart` enum, would that constitute a breaking API change? Since we're introducing a new enum variant. Maybe mark it as non-exhaustive, or would this be unnecessary since its mainly used as an input to the `date_part()` function?

Comment (tustvold) [reply arrow-arith/src/temporal.rs:384]: Non exhaustive sounds like a reasonable idea to me

Comment (Jefffrey) [reply arrow-arith/src/temporal.rs:384]: I'll add this and the deprecations for the shim functions, then PR should be good to go
